### PR TITLE
Selenium: Cover "Dashboard / Workspace Details / Machines" part of test plan by selenium test

### PR DESCRIPTION
**Pull request by Ohrimenko1988**
### What does this PR do?
Cover "Dashboard / Workspace Details / Machines" part of test plan by selenium test

### What issues does this PR fix or reference?
Issue: https://github.com/eclipse/che/issues/10692

### CODE REVIEW

**Review by musienko-maxim — APPROVED** [review]

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+    editMachineForm.waitNameInvalidHighliting();
+    editMachineForm.waitSaveButtonDisabling();
+  }
+
+  @Test
+  public void checkRamSection() {
+    // check machine name editing
+    workspaceDetailsMachines.clickOnEditButton(MACHINE_NAME);
+    editMachineForm.waitForm();
+    /*editMachineForm.typeName(CHANGED_MACHINE_NAME);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: Are really need to have commented code here?

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    final String machineListItemXpath = format(MACHINE_LIST_ITEM_XPATH_TEMPLATE, machineName);
+    return By.xpath(machineListItemXpath);
+  }
+
+  private By getCheckboxLocator(String machineName) {
+    final String checkboxXpath = format(CHECKBOX_XPATH_TEMPLATE, machineName);
+    return By.xpath(checkboxXpath);
+  }
+
+  private By getDecrementRamButtonLocator(String machineName) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java:104]: Too many simmilar getters here, IMHO. What about enclosing them into the enum?

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+  public WebElement waitSliderButton() {
+    return seleniumWebDriverHelper.waitVisibility(By.xpath(SLIDER_BUTTON_XPATH));
+  }
+
+  public WebElement waitSliderBar() {
+    return seleniumWebDriverHelper.waitVisibility(By.xpath(SLIDER_BAR_XPATH));
+  }
+
+  public void drugSliderButtonToSliderBarCenter() {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java]: "drug" > "drag"

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  public void typeRam(String ramValue) {
+    seleniumWebDriverHelper.setValue(waitRamTextField(), ramValue);
+  }
+
+  public void waitRamFieldText(String expectedText) {
+    seleniumWebDriverHelper.waitValueEqualsTo(waitRamTextField(), expectedText);
+  }
+
+  public boolean isRamValueValid() {
+    final String validationAttribute = "aria-invalid";
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java]: I would say that it is more likely 'invalidRamFlagAtribute'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    final String notificationMessageXpath =
+        format(NOTIFICATION_MESSAGE_XPATH_TEMPLATE, expectedMessage);
+    seleniumWebDriverHelper.waitVisibility(By.xpath(notificationMessageXpath));
+  }
+
+  public void waitMachinesListItem(String machineName) {
+    seleniumWebDriverHelper.waitVisibility(getMachineListItemLocator(machineName));
+  }
+
+  public void waitMachinesListItemWithAttributes(
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: 'MachinesList'  > 'MachineList'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+    return seleniumWebDriverHelper
+        .waitVisibilityAndGetAttribute(waitRamTextField(), validationAttribute)
+        .equals("false");
+  }
+
+  public void waitRamFieldValidHighlighting() {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java]: 'waitRamFieldValidHighlighting'  > 'waitValidRamFieldHighlighting'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+    seleniumWebDriverHelper.waitSuccessCondition(driver -> isRamValueValid());
+  }
+
+  public void waitRamFieldInvalidHighlighting() {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/EditMachineForm.java]: 'waitRamFieldInvalidHighlighting' > 'waitInvalideRamFieldHighlighting'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java [hunk 8]
```diff
@@ -140,6 +164,38 @@ public WorkspaceDetails(SeleniumWebDriver seleniumWebDriver, Loader loader, Dash
   @FindBy(xpath = Locators.CLOSE_DIALOG_BUTTON)
   WebElement closeBtn;
 
+  public WebElement wait(ActionButton actionButton) {
+    return seleniumWebDriverHelper.waitVisibility(actionButton.getLocator());
+  }
+
+  public void waitInvisibility(ActionButton... actionButtons) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java]: I would say that it waits **all** buttons invisibility

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java [hunk 9]
```diff
@@ -140,6 +164,38 @@ public WorkspaceDetails(SeleniumWebDriver seleniumWebDriver, Loader loader, Dash
...
+    waitState(actionButton.getLocator(), enabled);
+  }
+
+  private void waitState(By locator, boolean enabled) {
+    final String buttonStateAttribute = "aria-disabled";
+    seleniumWebDriverHelper.waitAttributeEqualsTo(
+        locator, buttonStateAttribute, Boolean.toString(!enabled));
+  }
+
+  public void waitDisabled(ActionButton... actionButtons) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java]: 'waitAllDisabled'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java [hunk 10]
```diff
@@ -140,6 +164,38 @@ public WorkspaceDetails(SeleniumWebDriver seleniumWebDriver, Loader loader, Dash
...
+    asList(actionButtons).forEach(actionButton -> waitState(actionButton, false));
+  }
+
+  public void waitEnabled(ActionButton... actionButtons) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetails.java]: 'waitAllEnabled'

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+            + "//ng-form[@name='ramAmountForm']//button[@data-action='increment']";
+    String RAM_AMOUNT_TEXT_FIELD_XPATH_TEMPLATE =
+        MACHINE_LIST_ITEM_XPATH_TEMPLATE + "//input[@id='machine--ram']";
+    String EDIT_BUTTON_XPATH_TEMPLATE = "//div[@edit-machine='%s']";
+    String SETTINGS_BUTTON_XPATH_TEMPLATE =
+        MACHINE_LIST_ITEM_XPATH_TEMPLATE + "//span[@uib-tooltip='Settings']";
+    String DELETE_BUTTON_XPATH_TEMPLATE = "//div[@delete-machine='%s']";
+    String SETTINGS_POPOVER_CONTAINER_XPATH =
+        "//div[contains(@class, 'workspace-machine-actions-popover') and @is-open='isOpen']";
+    String POPOVER_INSTALLERS_LINK_XPATH =
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: Could you, please, explain what does 'POPOVER' mean?

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+        .waitVisibilityAndGetAttribute(
+            getRamAmountTextFieldLocator(machineName), ramValidationAttribute)
+        .equals("false");
+  }
+
+  public void waitRamValidHighlighting(String machineName) {
+    seleniumWebDriverHelper.waitSuccessCondition(driver -> isRamValueValid(machineName));
+  }
+
+  public void waitRamInvalidHighlighting(String machineName) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: `waitRamInvalidHighlighting` > `waitInvalideRamHighlighting`

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+
+  public boolean isRamValueValid(String machineName) {
+    final String ramValidationAttribute = "aria-invalid";
+    return seleniumWebDriverHelper
+        .waitVisibilityAndGetAttribute(
+            getRamAmountTextFieldLocator(machineName), ramValidationAttribute)
+        .equals("false");
+  }
+
+  public void waitRamValidHighlighting(String machineName) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: `waitRamValidHighlighting` > `waitValidRamHighlighting`

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+
+    // close form by "Close" button
+    workspaceDetailsMachines.clickOnEditButton(MACHINE_NAME);
+    editMachineForm.waitForm();
+    editMachineForm.clickOnCloseIcon();
+    editMachineForm.waitFormInvisibility();
+  }
+
+  @Test
+  public void checkEditFormNameField() {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: IMHO test name `checkEditOfMachineName` is more suitable here.

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 15]
```diff
@@ -0,0 +1,267 @@
...
+    // check default values
+    workspaceDetailsMachines.clickOnEditButton(MACHINE_NAME);
+    editMachineForm.waitForm();
+    editMachineForm.waitName(MACHINE_NAME);
+    editMachineForm.waitSliderRamValue("2 GB");
+    editMachineForm.waitRamFieldText("2");
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: `"2"` > `EXPECTED_RAM_VALUE`

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 16]
```diff
@@ -0,0 +1,267 @@
...
+  private static final String MACHINE_NAME = "dev-machine";
+  private static final String CHANGED_MACHINE_NAME = "test-machine";
+  private static final String IMAGE_NAME = "eclipse/ubuntu_jdk8";
+  private static final String EXPECTED_RAM_VALUE = "2";
+  private static final String EMPTY_NAME_ERROR_MESSAGE = "Machine's name is required.";
+  private static final String SPECIAL_CHARACTERS_ERRORS_MESSAGE =
+      "The name should not contain special characters like space, dollar, etc.";
+  private static final String NAME_WITH_SPECIAL_CHARACTERS = "@#$^&*(!";
+  private static final String MAX_VALID_NAME = NameGenerator.generate("max_name", 120);
+  private static final String INVALID_LENGTH_NAME = NameGenerator.generate(MAX_VALID_NAME, 1);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: too small length is invalid as well.  So more straightforward name explain the value of constant clearer: `TOO_BIG_NAME`

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 17]
```diff
@@ -0,0 +1,267 @@
...
+  private static final String BIGGER_THAN_VALID_RAM_SIZE = "1000";
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: Simpler constant name: `TOO_BIG_RAM_SIZE`

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+    workspaceDetailsMachines.waitMachinesListItemWithAttributes(
+        MACHINE_NAME, IMAGE_NAME, CHANGED_RAM_SIZE);
+    workspaceDetails.waitEnabled(SAVE_BUTTON, APPLY_BUTTON, CANCEL_BUTTON);
+    workspaceDetails.waitAndClickOn(SAVE_BUTTON);
+    workspaceDetailsMachines.waitMachinesListItemWithAttributes(
+        MACHINE_NAME, IMAGE_NAME, CHANGED_RAM_SIZE);
+  }
+
+  @Test
+  public void checkSettingsButton() {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachineActionsTest.java]: Looks like the test is checking not just settings button. Therefore the test name `checkMachineSettings` would be  more suitable here.

File: selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachinesRamTest.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+    workspaceDetailsMachines.waitRamValidHighlighting(MACHINE_NAME);
+
+    workspaceDetailsMachines.clickOnIncrementRamButton(MACHINE_NAME);
+    workspaceDetailsMachines.waitRamAmount(MACHINE_NAME, "1");
+
+    workspaceDetailsMachines.typeRamAmount(MACHINE_NAME, "3");
+    workspaceDetailsMachines.waitRamValidHighlighting(MACHINE_NAME);
+    workspaceDetails.waitEnabled(SAVE_BUTTON);
+    workspaceDetails.waitAndClickOn(SAVE_BUTTON);
+    workspaceDetailsMachines.clickOnIncrementRamButton(MACHINE_NAME);
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/test/java/org/eclipse/che/selenium/dashboard/workspaces/details/WorkspaceDetailsMachinesRamTest.java]: Empty line between the steps was missed.

**Comment by dmytro-ndp** [comment]
ci-test

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/577/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/577//Selenium_20tests_20report/)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/10739)
docker image: eclipseche/che-server:10739
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by dmytro-ndp** [comment]
Please,  take a look at failed test https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/577/Selenium_20tests_20report/failsafe-report.html#org.eclipse.che.selenium.dashboard.workspaces.detailsWorkspaceDetailsMachinesRamTest

File: selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java [hunk 20]
```diff
@@ -193,8 +194,29 @@ public void waitRamAmount(String machineName, String ramAmount) {
...
+    seleniumWebDriverHelper.moveCursorTo(getEditButtonLocator(machineName));
+  }
+
   public void typeRamAmount(String machineName, String ramAmount) {
-    seleniumWebDriverHelper.setValue(getRamAmountTextFieldLocator(machineName), ramAmount);
+    moveCursorToEditButton(machineName);
+    clickOnRamField(machineName);
+    seleniumWebDriverHelper.setValue(getRamAmountTextFieldLocator(machineName), "");
+
+    if (ramAmount.isEmpty()) {
```
Comment (dmytro-ndp) [inline selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: It makes sense to check it at start of method to return as fast as possible.

Comment (Ohrimenko1988) [reply selenium/che-selenium-test/src/main/java/org/eclipse/che/selenium/pageobject/dashboard/workspaces/WorkspaceDetailsMachines.java]: Yes, for a case when specified empty string

**Comment by Ohrimenko1988** [comment]
ci-test

**Comment by dmytro-ndp** [comment]
ci-test

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/689/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/689//Selenium_20tests_20report/)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/10739)
docker image: eclipseche/che-server:10739
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Comment by Ohrimenko1988** [comment]
ci-test

**Review by dmytro-ndp — APPROVED** [review]
Adding selenium tests have passed here https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/695 according to the intermediate results.

**Comment by riuvshin** [comment]
ci-test build report:
[Build details](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/695/)
[Test report](https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/695//Selenium_20tests_20report/)
[selenium tests report data](http://tests.codenvycorp.com/selenium_test_report_data/10739)
docker image: eclipseche/che-server:10739
https://github.com/orgs/eclipse/teams/eclipse-che-qa please check this report.

**Review by SkorikSergey — APPROVED** [review]

**Pull request merged**